# Module 02 Homework - Pandas
Using a dataset from _Wine Spectator_, a wine magazine, we will practice the material that we covered in notebook M02_Colab04 to M02_Colab06. This notebook will cover data transformation, grouping, and sorting using pandas.

Created by GLORIA DUO (018038020)

Last updated: 22 FEBRUARY 2026

In [1]:
csvurl="https://gist.githubusercontent.com/clairehq/" + \
        "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"
import pandas as pd
wine = pd.read_csv(csvurl)

**Data cleaning**  
Notice that the first column is redundant. Part of data analysis is cleaning and removing redundancy. How would you drop the redundant column inplace, that is overwrite the dataframe.

**Answer:**
In order to drop the redundant column inplace, we can use the pandas.DataFrame.drop function and pass in inplace=True.

In [4]:
# Display dataframe head:
wine.head()

,Unnamed: 0,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
2,2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,87,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm
3,3,US,"Pineapple rind, lemon pith and orange blossom ...",Reserve Late Harvest,87,13.0,Michigan,Lake Michigan Shore,NaN,Alexander Peartree,NaN,St. Julian 2013 Reserve Late Harvest Riesling ...,Riesling,St. Julian
4,4,US,"Much like the regular bottling from 2012, this...",Vintner's Reserve Wild Child Block,87,65.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Sweet Cheeks 2012 Vintner's Reserve Wild Child...,Pinot Noir,Sweet Cheeks


In [5]:
# Display dataframe tail:
wine.tail()

,Unnamed: 0,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
65494,65494,France,Made from young vines from the Vaulorent porti...,Fourchaume Premier Cru,90,45.0,Burgundy,Chablis,NaN,Roger Voss,@vossroger,William Fèvre 2005 Fourchaume Premier Cru (Ch...,Chardonnay,William Fèvre
65495,65495,Australia,"This is a big, fat, almost sweet-tasting Caber...",NaN,90,22.0,South Australia,McLaren Vale,NaN,Joe Czerwinski,@JoeCz,Tapestry 2005 Cabernet Sauvignon (McLaren Vale),Cabernet Sauvignon,Tapestry
65496,65496,US,"Much improved over the unripe 2005, Fritz's 20...",Estate,90,20.0,California,Dry Creek Valley,Sonoma,NaN,NaN,Fritz 2006 Estate Sauvignon Blanc (Dry Creek V...,Sauvignon Blanc,Fritz
65497,65497,US,This wine wears its 15.8% alcohol better than ...,Block 24,90,31.0,California,Napa Valley,Napa,NaN,NaN,Hendry 2004 Block 24 Primitivo (Napa Valley),Primitivo,Hendry
65498,65498,Spain,"A unique take on Manzanilla Sherry, which is o...",Manzanilla,90,10.0,Andalucia,Jerez,NaN,Michael Schachner,@wineschach,Bodegas Dios Baco S.L. NV Manzanilla Sherry (J...,Sherry,Bodegas Dios Baco S.L.


In [6]:
# Display dataframe columns:
wine.columns

Index(['Unnamed: 0', 'country', 'description', 'designation', 'points',
       'price', 'province', 'region_1', 'region_2', 'taster_name',
       'taster_twitter_handle', 'title', 'variety', 'winery'],
      dtype='object')

In [7]:
# Drop the first column (redundant):
wine.drop(columns='Unnamed: 0', inplace=True)
wine.columns

Index(['country', 'description', 'designation', 'points', 'price', 'province',
       'region_1', 'region_2', 'taster_name', 'taster_twitter_handle', 'title',
       'variety', 'winery'],
      dtype='object')

#### Question 1: ####  
What is the mean of the points column?

There are many ways to calculate the mean, we can use pandas.DataFrame.describe and manually search for the answer, or we can sum the column and calculate the mean.

In [9]:
# Option 1: pandas.DataFrame.describe (default only outputs information for numerical columns)
wine.describe()

,points,price
count,65499.000000,60829.000000
mean,88.434037,35.232932
std,3.030310,39.477858
min,80.000000,4.000000
25%,86.000000,17.000000
50%,88.000000,25.000000
75%,91.000000,42.000000
max,100.000000,2500.000000


In [11]:
# Option 2: Select column, then use pandas.DataFrame.mean
points_mean = wine['points'].mean()
print("Mean of the points column:", points_mean)

Mean of the points column: 88.43403716087269


**Answer:** The mean of the points column is 88.43403716087269...

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

Since we only count each country once, we're looking for the number of unique countries. We can use pandas.DataFrame.nunique.

In [12]:
# Count the number of unique countries:
number_unique_countries = wine['country'].nunique()
print("Number of unique countries in dataset:", number_unique_countries)

Number of unique countries in dataset: 41


**Answer:** There are 41 unique countries present in the dataset.

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

We can find the frequency of each country by first selecting the country names, then using pandas.DataFrame.value_counts(ascending=True).

In [21]:
# Select the country name column and then set ascending=True.
wine['country'].value_counts(ascending=True)

country
Bosnia and Herzegovina        1
Armenia                       1
Slovakia                      1
Switzerland                   4
Luxembourg                    4
India                         4
Ukraine                       5
Macedonia                     6
Czech Republic                6
Cyprus                        6
Serbia                        7
Peru                          8
Morocco                      11
Lebanon                      20
Moldova                      30
Brazil                       31
Mexico                       31
England                      36
Georgia                      37
Slovenia                     39
Turkey                       43
Croatia                      44
Uruguay                      61
Hungary                      61
Romania                      67
Bulgaria                     68
Canada                      108
Greece                      242
Israel                      259
New Zealand                 733
South Africa                737


#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [26]:
# 'adjusted price' = price - average price (centering transformation). 
# Not sure if this is meant to be adjusted by the average price of each wine type or all wines in general.
adjusted_price = wine['price'] - wine['price'].mean()
print("Adjusted prices of wine:", adjusted_price) # Note that resulting NaN is due to us not dropping missing values before calculation.

# Verification (mean should be 0 now that it's centered):
print("Mean adjusted price (expected 0 or very small number):", adjusted_price.mean())

Adjusted prices of wine: 0              NaN
1       -20.232932
2       -21.232932
3       -22.232932
4        29.767068
           ...    
65494     9.767068
65495   -13.232932
65496   -15.232932
65497    -4.232932
65498   -25.232932
Name: price, Length: 65499, dtype: float64
Mean adjusted price (expected 0 or very small number): -3.229559183661617e-15


#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

The points-to-price ratio can be calculuated by dividing the points value in each row by the price value in each row. We can find the highest by applying pandas.DataFrame.max to the points-to-price ratio column.

In [31]:
# Calculate the points-to-price ratio:
points_to_price_ratio = wine['points'] / wine['price']

# Locate the position of the maximum:
max_index = points_to_price_ratio.idxmax() 

# Query the name of the wine in the index of the maximum ratio:
highest_wine = wine['title'].iloc[max_index] # This only works when we haven't modified the index in any way (which we haven't... yet).
# More robust query:
highest_wine_robust = wine.loc[max_index, 'title']

print("Wine with the highest points-to-price ratio in dataset:", highest_wine)
print("Testing robustness:", highest_wine_robust)

Wine with the highest points-to-price ratio in dataset: Bandit NV Merlot (California)
Testing robustness: Bandit NV Merlot (California)


**Answer:** The title of the wine with the highest points-to-price ratio in the dataset is Bandit NV Merlot (California).

#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

We can use pandas.Series.str.contains to help filter for specific description keywords. More specifically, we need to first select the description column, then pass in the regex to match with case sensitivity set to false. We cannot use literal string matching because that will match substrings too. Just in case there is no description (NaN), we instruct the function to ignore it. Afterwards, we need to apply the pandas.Series.sum function to count the number of descriptions containing matching keywords.

In [35]:
# Look for the word "tart", not case sensitive, not a substring of another word.
# Same applies to the word "berries"
flavor_counts = pd.Series({
    "Tart": wine['description'].str.contains(r"\btart\b", case=False, na=False, regex=True).sum(), # regex is True by default.
    "Berries": wine['description'].str.contains(r"\bberries\b", case=False, na=False, regex=True).sum()
})

flavor_counts

Tart       3086
Berries    1192
dtype: int64

#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset.

We'll need to build the star rating columns step-by-step.

In [36]:
# Let's start with the default, 1 star:
wine['star_ratings'] = 1

# Now we can handle 2 star ratings using pandas.DataFrame.loc to select rows where our conditions are True, then select the given column.
wine.loc[wine['points'] >= 80, 'star_ratings'] = 2

# Repeat for other 3 stars.
wine.loc[wine['points'] >= 90, 'star_ratings'] = 3
wine.loc[wine['country'] == 'France', 'star_ratings'] = 3

wine.head()

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery,star_ratings
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia,2
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos,2
2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,87,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm,2
3,US,"Pineapple rind, lemon pith and orange blossom ...",Reserve Late Harvest,87,13.0,Michigan,Lake Michigan Shore,NaN,Alexander Peartree,NaN,St. Julian 2013 Reserve Late Harvest Riesling ...,Riesling,St. Julian,2
4,US,"Much like the regular bottling from 2012, this...",Vintner's Reserve Wild Child Block,87,65.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Sweet Cheeks 2012 Vintner's Reserve Wild Child...,Pinot Noir,Sweet Cheeks,2


#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

Since we're not renaming any indices or creating new ones, we don't need to specify the index name like for Question 6. We can just use pandas.DataFrame.value_counts for this problem.

In [39]:
# Create review_counts:
review_counts = wine['taster_twitter_handle'].value_counts()

print("Most common wine reviewers in dataset (Top 10):")
review_counts.head(10)

Most common wine reviewers in dataset (Top 10):


taster_twitter_handle
@vossroger          13045
@wineschach          7752
@kerinokeefe         5313
@paulgwine           4851
@vboone              4696
@mattkettmann        3035
@JoeCz               2605
@wawinereport        2358
@gordone_cellars     2032
@AnneInVino          1769
Name: count, dtype: int64

#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

This is another situation where pandas.DataFrame.value_counts can be used. Alternatively, we can create the multi-index using pandas.DataFrame.groupby, then pandas.Series.sort_values.

In [40]:
# Option 1: pandas.DataFrame.value_counts
country_variety_wine_counts = (wine[['country', 'variety']].value_counts())
country_variety_wine_counts.head(10)

country  variety                 
US       Pinot Noir                  4918
         Cabernet Sauvignon          3649
         Chardonnay                  3412
France   Bordeaux-style Red Blend    2380
Italy    Red Blend                   1870
US       Syrah                       1588
         Red Blend                   1459
France   Chardonnay                  1421
US       Zinfandel                   1362
Italy    Nebbiolo                    1340
Name: count, dtype: int64

In [42]:
# Option 2: pandas.DataFrame.groupby into pandas.DataFrameGroupBy.size into pandas.Series.sort_values
country_variety_wine_counts_alternative = (wine.groupby(['country', 'variety']).size().sort_values(ascending=False))
country_variety_wine_counts_alternative.head(10)

country  variety                 
US       Pinot Noir                  4918
         Cabernet Sauvignon          3649
         Chardonnay                  3412
France   Bordeaux-style Red Blend    2380
Italy    Red Blend                   1870
US       Syrah                       1588
         Red Blend                   1459
France   Chardonnay                  1421
US       Zinfandel                   1362
Italy    Nebbiolo                    1340
dtype: int64

#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer.  
*Hint:* You will need the `taster_name` and `points` columns.

In [43]:
# We can use pandas.DataFrame.groupby on the reviewer name, then select the points column for each group, then calculate their average score using pandas.DataFrameGroupBy.mean.
reviewer_average_scores = wine.groupby('taster_name')['points'].mean()
reviewer_average_scores.head(10)

taster_name
Alexander Peartree    86.014286
Anna Lee C. Iijima    88.380506
Anne Krebiehl MW      90.587903
Carrie Dykes          86.644444
Christina Pickard     89.500000
Fiona Adams           87.090909
Jeff Jenssen          88.273504
Jim Gordon            88.604331
Joe Czerwinski        88.519770
Kerin O’Keefe         88.827969
Name: points, dtype: float64